activesupport/lib/active_support/cache/key.rb

```diff
@@ -0,0 +1,153 @@
+    #   # => true
+    #
+    # In addition to passing a key into the initialize method,
+    # key fragments can be added onto an existing key through the
+    # `<<` method.
+    #
+    # Example:
+    #
+    #   key = Key.new
+    #   key << "foo"
+    #   key << "bar"
+    #   key.cache_key
+    #   # => "foo/bar"
+    #
+    # Multiple key entries generate the same key regardless of how
+    # they are added. I.e. if they're added via multiple arrays
+    # or directly via the `<<` operator.
+    #
+    # Example:
+    #
+    #   k1 = Key.new("foo")
+    #   k1 << "bar"
+    #   k2 = Key.new(["foo", "bar"])
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #   k3 = Key.new(["foo", ["bar"]])
+    #   k2.cache_key == k3.cache_key
+    #   # => true
+    #
+    # It is faster to directly add objects to the
+    # Key object via `<<` than to generate
+    # an intermediate array of objects.
+    #
+    # The class supports strings, arrays, hashes,
+    # objects with a `cache_key` method and anything
+    # that can be implicitly cast to a string:
+    #
+    # Examples:
+    #
+    #   Key.new("foo").cache_key
+    #   # => "foo"
+    #
+    #   Key.new(["foo", "bar"]).cache_key
+    #   # => "foo/bar"
+    #
+    #   Key.new({ bar: 1, foo: 2 }).cache_key
+    #   # => "bar=1/foo=2"
+    #
+    #   klass = Class.new { def cache_key; "foo"; end }
+    #   Key.new(klass.new).cache_key
+    #   # => "foo"
+    #
+    # In addition to generating a `cache_key`,
+    # a `cache_version` is also generated if the object
+    # passed in responds to `cache_version` or if it contains
+    # an element that responds to `cache_version`.
+    #
+    # The purpose of keeping this value seperate from
+    # `cache_key` is to support the use of "cache versioning"
+    # where the same key is recycled but the version of the object
+    # being stored is kept directly in the cache.
+    #
+    # Example:
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new(klass.new).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new([klass.new]).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new({foo: klass.new}).cache_version
+    #   # => "foo"
+    class Key # :nodoc:
+      DEFAULT_KEY = Object.new
+      CACHE_METHOD_OBJ = ActiveSupport::Cache::Store.new
+      @@has_rails_cache = :unknown
+
+      def initialize(key = DEFAULT_KEY)
+        @cache_key = nil
+        @cache_version = nil
+        @key_parts = []
+        @cache_method_obj = cache_method_obj
+
+        self << key unless key == DEFAULT_KEY
+      end
+
+      def cache_key
+        @cache_key ||= @cache_method_obj.send(:expanded_key, @key_parts)
+      end
+
+      def cache_version
+        @cache_version ||= @cache_method_obj.send(:expanded_version, @key_parts)
+
+        if @cache_version == "/"
+          nil
```

Comment:
Without this code we get three failures in the tests:

```
CachedCollectionViewRenderTest#test_collection_caching_with_partial_that_doesn't_use_fragment_caching [/Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:797]:
Expected: "Cached"
  Actual: "Hello: david"


rails test Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:791

.................................F

Failure:
CachedCollectionViewRenderTest#test_collection_caching_with_cached_true [/Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:807]:
--- expected
+++ actual
@@ -1 +1,2 @@
-"Cached"
+"  Hello: david
+"



rails test Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:801

..F

Failure:
CachedCollectionViewRenderTest#test_template_body_written_to_cache [/Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:778]:
Expected: "Hello: david"
  Actual: nil


rails test Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:773
```

Here's an example of one of the tests:

```ruby
  test "collection caching with cached true" do
    customer = CachedCustomer.new("david", 1)
    key = cache_key(customer, "test/_cached_customer")

    ActionView::PartialRenderer.collection_cache.write(key, "Cached")

    assert_equal "Cached",
      @view.render(partial: "test/cached_customer", collection: [customer], cached: true)
  end
```

The issue at heart is that `nil` and empty values weren't really taken into account when the current `noramlize_version` was written so the behavior is somewhat undefined (or rather it's unclear what behavior we _want_).

For example if you try to generate a cache version based off of a symbol you'll get a nil, but if you try to generate it off of a symbol in an array you'll get an empty string:

```ruby
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, :views)
=> nil
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, [:views])
=> ""
```

These tests are failing because of the base assumption that normalize version will treat `[key]` the same as `key`. This is mostly true, but not in the case of empty values. In these failures we're generating a cache_version when we wouldn't have originally

Basically the difference between:

```
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, [[:views], "value", []])
# => "/"
```

and 

```
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, [:views, "value", []])
# => ""
```

My initial solution was to consider "/" the same as an empty string. I'm looking into seeing if there's another way to generate or insert keys that doesn't require this explicit knowledge.

Reply:
I've got a solution here https://github.com/rails/rails/pull/37301/commits/a01a94568cd38ceb40bd27d5ef8771565607b193

Essentially we can avoid this issue by making sure that items are directly added to the `Cache::Key`, so rather than adding a whole array to the key, we can add it's constituent parts. This method has the benefit of removing an array allocation from the internal loop.